src/ukify/ukify.py

```diff
@@ -128,6 +129,43 @@ def try_import(modname, name=None):
     except ImportError as e:
         raise ValueError(f'Kernel is compressed with {name or modname}, but module unavailable') from e
 
+def get_zboot_kernel(f):
+    """Decompress zboot efistub kernel if compressed. Return contents."""
+    # See linux/drivers/firmware/efi/libstub/Makefile.zboot
+    # and linux/drivers/firmware/efi/libstub/zboot-header.S
+
+    # Reading 4 bytes from address 0x08 is the starting offset of compressed data
+    f.seek(8)
+    _start = f.read(4)
+    start = struct.unpack('<i', _start)[0]
+
+    # Reading 4 bytes from address 0x0c is the size of compressed data,
+    # but it needs to be corrected according to the compressed type.
+    f.seek(0xc)
+    _sizes = f.read(4)
+    size = struct.unpack('<i', _sizes)[0]
+
+    # Read 6 bytes from address 0x18, which is a nul-terminated
+    # string representing the compressed type.
+    f.seek(0x18)
+    comp_type = f.read(6)
+    f.seek(start)
+    if comp_type.startswith(b'gzip'):
+        gzip = try_import('gzip')
+        return gzip.open(f).read(size)
+    if comp_type.startswith(b'lz4'):
+        lz4 = try_import('lz4.frame', 'lz4')
+        return lz4.frame.decompress(f.read(size))
+    if comp_type.startswith(b'lzma'):
+        lzma = try_import('lzma')
+        return lzma.open(f).read(size)
+    if comp_type.startswith(b'lzo'):
+        raise NotImplementedError('lzo decompression not implemented')
+    if comp_type.startswith(b'xzkern'):
+        raise NotImplementedError('xzkern decompression not implemented')
+    if comp_type.startswith(b'zstd22'):
+        zstd = try_import('zstd')
+        return zstd.uncompress(f.read(size))
```

Comment:
hmm, there should probably some logic here for handling unexpected values. I mean, my educated guess is that sooner or later more will addded here, and righ tnow we'll just return None here, which doesn't seem right.

maybe get rid of the explicit lzo/xzkern handling here, and just have a single generic fallback for all unknown and unexpected types that bails out with a an error showing the header string.

Reply:
I think the explicit handling of those is fine… It gives a clear message to the user and also makes it very simple to add support for those compressions is somebody wants to. But yeah, we shouldn't return None if nothing matches. Please throw an exception.